This GitHub repository contains course material for GMS7930: ChIP-seq data analysis practice.

We use Rmarkdown in this class to demonstrate data analysis modules.
In this class, ChIP-seq sequencing reads for demonstration are
stored as [BAM](https://en.wikipedia.org/wiki/Binary_Alignment_Map)
files in the
[**data**](https://github.com/tengmx/gms7930/tree/master/data)
folder, while the Rmarkdown file (*gms7930-rgs.Rmd*) and
its output file (*gms7930-rgs.html*) are located in the repository root folder.

To open and run the Rmarkdown file,
 [*rstudio*](https://posit.co/download/rstudio-desktop/)
 has to be installed firstly. 
Make sure you have the latest R release version installed in your computer.
Also, if you are running rstudio on Windows, make sure you run it as system administrator.

After all set above, you are ready to explore the data analysis modules documented in
the *gms7930-rgs.Rmd* file. Details of module instruction and function are embedded
inside the *gms7930-rgs.Rmd* file.

If you find it hard to read/edit the Rmarkdown file, you can focus on the html file
 (*gms7930-rgs.html*). To view the html file, you will need to download it to your computer,
then open it with common web browser such as Chrome and Firefox.
   
